cardinal/system/system_manager.go

```diff
@@ -0,0 +1,149 @@
+package system
+
+import (
+	"fmt"
+	"github.com/rotisserie/eris"
+	"path/filepath"
+	"pkg.world.dev/world-engine/cardinal/statsd"
+	"pkg.world.dev/world-engine/cardinal/types/engine"
+	"reflect"
+	"runtime"
+	"slices"
+	"time"
+)
+
+type Manager struct {
+	// registeredSystems is a list of all the registered system names in the order that they were registered.
+	// This is represented as a list as maps in Go are unordered.
+	registeredSystems []string
+
+	// registeredInitSystems is a list of all the registered init system names in the order that they were registered.
+	// This is represented as a list as maps in Go are unordered.
+	registeredInitSystems []string
+
+	// systemFn is a map of system names to system functions.
+	systemFn map[string]System
+
+	// currentSystem is the name of the system that is currently running.
+	currentSystem *string
+}
+
+// NewManager creates a new system manager.
+func NewManager() *Manager {
+	return &Manager{
+		registeredSystems: make([]string, 0),
+		systemFn:          make(map[string]System),
+		currentSystem:     nil,
+	}
+}
+
+// RegisterSystems registers multiple systems with the system manager.
+// There can only be one system with a given name, which is derived from the function name.
+// If there is a duplicate system name, an error will be returned and none of the systems will be registered.
+func (m *Manager) RegisterSystems(systems ...System) error {
+	return m.registerSystems(systems...)
+}
+
+// RegisterInitSystems registers multiple init systems that is only executed once at tick 0 with the system manager.
+// There can only be one system with a given name, which is derived from the function name.
+// If there is a duplicate system name, an error will be returned and none of the systems will be registered.
+func (m *Manager) RegisterInitSystems(systems ...System) error {
+	return m.registerSystems(systems...)
+}
+
+func (m *Manager) registerSystems(systems ...System) error {
+	// Iterate through all the systems and check if they are already registered.
+	// This is done before registering any of the systems to ensure that all are registered or none of them are.
```

Comment:
I don't understand why it's important to guarantee that either all or none of the systems are registered. This guarantee is also made in the message manager. 

Errors here will likely end up as a log.Fatal. Do you have a use case in mind where the user would depend on this all-or-nothing functionality? 

If it's not needed, this method can be simplified to a single loop over `systems`, no extra systemNames slice, and no slices.Contains call. (same goes for RegisterMessages)

Reply:
It's just defensive programming. I prefer that my program is never in an invalid state without making assumptions of external error handling.